movai_core_shared/core/zmq_server.py

```diff
@@ -0,0 +1,166 @@
+"""
+   Copyright (C) Mov.ai  - All Rights Reserved
+   Unauthorized copying of this file, via any medium is strictly prohibited
+   Proprietary and confidential
+
+   Usage:
+        Basic 0MQ client for connecting 0MQ servers.
+
+   Developers:
+   - Erez Zomer (erez@mov.ai) - 2023
+"""
+from abc import ABC, abstractmethod
+import asyncio
+import sys
+import json
+import logging
+import zmq
+import zmq.asyncio
+
+from movai_core_shared.exceptions import UnknownRequestError
+from movai_core_shared.messages.general_data import Request
+
+    
+class ZMQServer(ABC):
+    """
+    This class is a base class for any ZMQ server.
+    """
+    def __init__(self, server_name: str, bind_addr: str, new_loop: bool = False, debug: bool = False) -> None:
+        """Constructor"""
+        if not isinstance(server_name, str):
+            raise ValueError("server name must be of type string.")
+        if not isinstance(bind_addr, str):
+            raise ValueError("bind address must be of type string.")
+        if not isinstance(new_loop, bool):
+            raise ValueError("new_loop must be of type bool.")
+        if not isinstance(debug, bool):
+            raise ValueError("debug must be of type bool.")
+        self._name = server_name
+        self._addr = bind_addr
+        self._logger = logging.getLogger(server_name)
+        self._new_loop = new_loop
+        self._debug = debug
+        self._initialized = False
```

Comment:
maybe initialized should be class arg
and a dictionary 
so if you wanna start another server 
with the same address
it will check the class for previous server that were initialized with the same address

Reply:
this class binds to an address so if you initialize another object the address must be different unless an exception will be raised

Reply:
but the bind is not on the constructor 
so you can 2 objects
that one is connected 
and one is not
and if you get an exception 
you will need to delete the one that isn't connected

Reply:
2 servers can coexist on the same ip, it's the port that makes the difference...